src/main/java/net/minecraftforge/event/world/JigsawPieceEvent.java

```diff
@@ -0,0 +1,57 @@
+package net.minecraftforge.event.world;
+
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.gen.feature.jigsaw.JigsawPiece;
+import net.minecraftforge.eventbus.api.Event;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+/**
+ * Is fired from {@link net.minecraft.world.gen.feature.jigsaw.JigsawManager.Assembler#tryPlacingChildren}
+ * for each generated Jigsaw block after it stores the pool of Jigsaw pieces in can attach to.
+ * The list does NOT contain the pieces contained in {@link net.minecraft.world.gen.feature.jigsaw.JigsawPattern#fallback}
+ * to avoid infinite generation problems.
+ * <br><br>
+ * Adding a custom villager's house to plains village and zombie village can be just:
+ *
+ * {@code
+ *    String pool = event.patternLocation.getPath();
+ *    if (pool.startsWith("village/plains/houses") || pool.startsWith("village/plains/zombie/houses"))
+ *      event.pieces.add(new SingleJigsawPiece("namespace:path"));
+ * }
+ *
+ * For more detailed insight see <a href=https://mcforge.readthedocs.io/en/1.16.x/events/jigsawevent/>here.</a>
+ */
+public class JigsawPieceEvent extends Event
+{
+   public List<JigsawPiece> pieces;
+   public ResourceLocation patternLocation;
+   public BlockPos jigsawBlockPosition;
+   public Map<String, Object> jigsawAssemblerSettings;
```

Comment:
So it would be `public final Map<Resourcelocation, Object> jigsawAssemblerSettings;`? But will that allow arbitrary keys such as `villagers` and `generated` (that I used in test) or other settings such as `targetEntranceY` (something like what I mentioned on forums)?

Reply:
yes, then you must use `modid:generated`, because another mod could also use `generated` and with the modid it should be unique.

Reply:
Ok, I will use `Resourcelocation`. With `string` I though any real mod would use something like `modname_structurename_settingname` but I guess it will be more clear with `Resourcelocation`.